proposals/nnnn-enhance-video-streaming-performance-for-android.md

```diff
@@ -0,0 +1,115 @@
+
+Due to the fact that we have tested, the video streaming performance is enhanced.
+
+For `receive()` array function:
+```java
+/**
+ * Receives an array of bytes
+ * @param bytes data
+ * @param byteOffset offset of the data
+ * @param byteCount count of data
+ * @throws IOException
+ */
+synchronized void receive(byte[] bytes, int byteOffset, int byteCount) throws IOException {
+    checkStateForReceive();
+    this.writeSide = Thread.currentThread();
+    int bytesToTransfer = byteCount;
+    while (bytesToTransfer > 0) {
+        if (this.in == this.out) {
+            awaitSpace();
+        }
+        int nextTransferBytes = 0;
+        if (this.out < this.in) {
+            nextTransferBytes = this.buffer.length - this.in;
+        } else if (this.in < this.out) {
+            if (this.in == -1) {
+                this.in = this.out = 0;
+                nextTransferBytes = this.buffer.length - this.in;
+            } else {
+                nextTransferBytes = this.out - this.in;
+            }
+        }
+        if (nextTransferBytes > bytesToTransfer) {
+            nextTransferBytes = bytesToTransfer;
+        }
+        System.arraycopy(bytes, byteOffset, this.buffer, this.in, nextTransferBytes);
+        bytesToTransfer -= nextTransferBytes;
+        byteOffset += nextTransferBytes;
+        this.in += nextTransferBytes;
+        if (this.in >= this.buffer.length) {
+            this.in = 0;
+        }
+    }
+    notifyAll();
+}
+```
+
+For `write()` array function:
+```java
+/**
+ * Write an array of bytes
+ * @param bytes array of data
+ * @param byteOffset offset of array
+ * @param byteCount count of data
+ * @throws IOException
+ */
+public void write(byte[] bytes, int byteOffset, int byteCount) throws IOException {
+    if (this.sink == null) {
+        throw new IOException("Pipe not connected");
+    }else if (bytes == null) {
+        throw new NullPointerException();
+    }else if ((byteOffset < 0) || (byteOffset > bytes.length) || (byteCount < 0)
+            || (byteOffset + byteCount > bytes.length) || (byteOffset + byteCount < 0)) {
+        throw new IndexOutOfBoundsException();
+    }else if (byteCount == 0) {
+        return;
+    }
+    this.sink.receive(bytes, byteOffset, byteCount);
+}
+
+```
+
+## Potential downsides
+
+N/A
+
+## Impact on existing code
+
+There is no impact for the existing code. The changes in this proposal is to remove the default io PipedInputStream and PipedOutputStream packge and reimport PipedInputStream and PipedOutputStream from component streaming. The developer would not need to change their previous code if they do not wish to use the new library for video streaming.
+ 
+## Alternatives considered
+
+There is a optional solution to improve the video streaming performance is to straightly send data from encoder byte buffer and handler to avoid using the pipe stream.
```

Comment:
Please update to "There is an optional solution to improve the video streaming performance by directly sending data..."